# 단순 회귀분석

In [1]:
import pandas as pd
from statsmodels.formula.api import ols

In [2]:
df = pd.read_csv('../datasets/iris.csv')
df.head(2)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

In [3]:
# model = ols(formula="Sepal.Length ~ Sepal.Width", data=df).fit()
# model

PatsyError: Error evaluating factor: NameError: name 'Sepal' is not defined
    Sepal.Length ~ Sepal.Width
                   ^^^^^^^^^^^

In [4]:
df.columns = ["SL", "SW", "PL", "PW", "species"]
df.head(2)

SL   SW   PL   PW species
0  5.1  3.5  1.4  0.2  setosa
1  4.9  3.0  1.4  0.2  setosa

In [5]:
model = ols(formula="SL ~ SW", data=df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     SL   R-squared:                       0.014
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     2.074
Date:                Mon, 14 Nov 2022   Prob (F-statistic):              0.152
Time:                        15:25:48   Log-Likelihood:                -183.00
No. Observations:                 150   AIC:                             370.0
Df Residuals:                     148   BIC:                             376.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.5262      0.479     13.628      0.000       5.580       7.473
SW            -0.2234      0.155     -1.440      0.152      -0.530       0.083
==============================================================================
Omnibus:                        4.389   Durbin-Watson:                   0.952
Prob(Omnibus):                  0.111   Jarque-Bera (JB):                4.237
Skew:                           0.360   Prob(JB):                        0.120
Kurtosis:                       2.600   Cond. No.                         24.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [7]:
model = ols(formula="PL ~ PW", data=df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     PL   R-squared:                       0.927
Model:                            OLS   Adj. R-squared:                  0.927
Method:                 Least Squares   F-statistic:                     1882.
Date:                Mon, 14 Nov 2022   Prob (F-statistic):           4.68e-86
Time:                        15:27:55   Log-Likelihood:                -101.18
No. Observations:                 150   AIC:                             206.4
Df Residuals:                     148   BIC:                             212.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.0836      0.073     14.850      0.000       0.939       1.228
PW             2.2299      0.051     43.387      0.000       2.128       2.332
==============================================================================
Omnibus:                        2.438   Durbin-Watson:                   1.430
Prob(Omnibus):                  0.295   Jarque-Bera (JB):                1.966
Skew:                           0.211   Prob(JB):                        0.374
Kurtosis:                       3.369   Cond. No.                         3.70
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [9]:
model.predict(df.iloc[:6, ])

0    1.529546
1    1.529546
2    1.529546
3    1.529546
4    1.529546
5    1.975534
dtype: float64

In [11]:
df['pred'] = model.predict(df)
df.head()

SL   SW   PL   PW species      pred
0  5.1  3.5  1.4  0.2  setosa  1.529546
1  4.9  3.0  1.4  0.2  setosa  1.529546
2  4.7  3.2  1.3  0.2  setosa  1.529546
3  4.6  3.1  1.5  0.2  setosa  1.529546
4  5.0  3.6  1.4  0.2  setosa  1.529546

In [12]:
from sklearn.linear_model import LinearRegression

In [13]:
model = LinearRegression().fit(X=df[['PL']], y=df[['PW']])  # 이중 대괄호는 데이터프레임이 풀리지 않음

In [14]:
model.coef_

array([[0.41575542]])

In [15]:
model.intercept_

array([-0.36307552])

In [16]:
model.predict(df[['PL']])

array([[0.21898206],
       [0.21898206],
       [0.17740652],
       [0.2605576 ],
       [0.21898206],
       [0.34370869],
       [0.21898206],
       [0.2605576 ],
       [0.21898206],
       [0.2605576 ],
       [0.2605576 ],
       [0.30213314],
       [0.21898206],
       [0.09425544],
       [0.13583098],
       [0.2605576 ],
       [0.17740652],
       [0.21898206],
       [0.34370869],
       [0.2605576 ],
       [0.34370869],
       [0.2605576 ],
       [0.0526799 ],
       [0.34370869],
       [0.42685977],
       [0.30213314],
       [0.30213314],
       [0.2605576 ],
       [0.21898206],
       [0.30213314],
       [0.30213314],
       [0.2605576 ],
       [0.2605576 ],
       [0.21898206],
       [0.2605576 ],
       [0.13583098],
       [0.17740652],
       [0.21898206],
       [0.17740652],
       [0.2605576 ],
       [0.17740652],
       [0.17740652],
       [0.17740652],
       [0.30213314],
       [0.42685977],
       [0.21898206],
       [0.30213314],
       [0.218

In [17]:
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error

In [18]:
mean_absolute_error(y_true=df['PL'], y_pred=df['PW'])

2.558666666666667

In [19]:
mean_squared_error(y_true=df['PL'], y_pred=df['PW'])

7.645466666666667

In [20]:
mean_squared_error(y_true=df['PL'], y_pred=df['PW']) ** 0.5  # RMSE(Root Mean Squared Error, 0.5제곱은 제곱근임)

2.76504370067937

### 1. 종속변수를 registered, 독립변수를 temp로 했을 때 결정계수는?

bike.csv 파일 사용
statsmodels 함수 사용
학습 데이터 비율을 70%, seed를 123으로 설정

1. 0.106
2. 0.101
3. 0.152
4. 0.116

In [21]:
df = pd.read_csv('../datasets/bike.csv')
df.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [22]:
from sklearn.model_selection import train_test_split

In [23]:
df_train, df_test = train_test_split(df, train_size=0.7, random_state=123)
print(df_train.head(2))
print(df_test.head(2))

                 datetime  season  holiday  workingday  weather   temp  \
4046  2011-09-19 15:00:00       3        0           1        2  24.60   
9262  2012-09-09 07:00:00       3        0           0        1  22.14   

       atemp  humidity  windspeed  casual  registered  count  
4046  30.305        60    15.0013      44         143    187  
9262  25.760        73    11.0014      20          50     70  
                 datetime  season  holiday  workingday  weather   temp  atemp  \
6495  2012-03-07 21:00:00       1        0           1        1  18.04  21.97   
7050  2012-04-12 03:00:00       2        0           1        1  13.12  15.15   

      humidity  windspeed  casual  registered  count  
6495        62    16.9979      11         177    188  
7050        61    16.9979       2           3      5  


In [24]:
model = ols(formula='registered ~ temp', data=df_train).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             registered   R-squared:                       0.106
Model:                            OLS   Adj. R-squared:                  0.106
Method:                 Least Squares   F-statistic:                     902.3
Date:                Mon, 14 Nov 2022   Prob (F-statistic):          1.92e-187
Time:                        15:51:05   Log-Likelihood:                -48650.
No. Observations:                7620   AIC:                         9.730e+04
Df Residuals:                    7618   BIC:                         9.732e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     27.5151      4.559      6.036      0.000      18.579      36.452
temp           6.3391      0.211     30.038      0.000       5.925       6.753
==============================================================================
Omnibus:                     2097.525   Durbin-Watson:                   2.022
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5337.402
Skew:                           1.502   Prob(JB):                         0.00
Kurtosis:                       5.790   Cond. No.                         60.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 2. 종속변수를 `casual`, 독립변수를 atemp로 했을 때 RMSE는?

bike.csv 파일 사용
statsmodels 함수 활용
학습 데이터 비율을 70%, seed를 123으로 설정

1. 1957.2
2. 44.2
3. 23.3
4. 44.5

In [25]:
df = pd.read_csv('../datasets/bike.csv')
df_train, df_test = train_test_split(df, train_size=0.7, random_state=123)

In [26]:
model = ols(formula='casual ~ atemp', data=df_train).fit()
pred = model.predict(df_test)
pred[:4]

C:\Users\Lee\AppData\Local\Temp\ipykernel_4840\4112327965.py:3: FutureWarning: The behavior of `series[i:j]` with an integer-dtype index is deprecated. In a future version, this will be treated as *label-based* indexing, consistent with e.g. `series[i]` lookups. To retain the old behavior, use `series.iloc[i:j]`. To get the future behavior, use `series.loc[i:j]`.
  pred[:4]


6495    31.499001
7050    12.626390
558     10.537120
5085    33.588271
dtype: float64

In [27]:
mean_squared_error(y_pred=pred, y_true=df_test['casual']) ** 0.5

44.46237010271433

### 3. 종속변수를 casual, 독립변수를 atemp로 했을 때 여름과 겨울의 RMSE 차이는?

bike.csv 파일 사용
statsmodels 함수 활용
학습 데이터 비율을 70%, seed를 123으로 설정
RMSE의 차이는 절대값을 취한다.

1. 41.1
2. 49.8
3. 8.6
4. 12.4

In [40]:
df = pd.read_csv('../datasets/bike.csv')
df_s2 = df.loc[df['season'] == 2,]
df_s4 = df.loc[df['season'] == 4,]

In [41]:
df_s2_train, df_s2_test = train_test_split(df_s2, train_size=0.7, random_state=123)
df_s4_train, df_s4_test = train_test_split(df_s4, train_size=0.7, random_state=123)

In [43]:
model_s2 = ols(formula='casual ~ atemp', data=df_s2_train).fit()
model_s4 = ols(formula='casual ~ atemp', data=df_s4_train).fit()
pred_s2 = model_s2.predict(df_s2_test)
pred_s4 = model_s4.predict(df_s4_test)

In [46]:
RMSE_s2 = mean_squared_error(y_pred=pred_s2, y_true=df_s2_test['casual']) ** 0.5
RMSE_s4 = mean_squared_error(y_pred=pred_s4, y_true=df_s4_test['casual']) ** 0.5
abs(RMSE_s2 - RMSE_s4)

8.648423450414171